quodlibet/qltk/entry.py

```diff
@@ -238,30 +239,46 @@ class ClearEntry(UndoEntry, ClearEntryMixin):
     __gsignals__ = ClearEntryMixin.__gsignals__
 
 
-class ValidatingEntryMixin:
+class ValidatingEntryMixin(Gtk.Widget):
     """An entry with visual feedback as to whether it is valid or not.
     The given validator function gets a string and returns True (green),
     False (red), or None (black).
     """
 
-    INVALID = Gdk.RGBA(0.8, 0, 0)
-    VALID = Gdk.RGBA(0.3, 0.6, 0.023)
+    ALPHA = 1.0
+    INVALID = Gdk.RGBA(0.85, 0.0, 0.05, ALPHA)
+    VALID = Gdk.RGBA(0.0, 0.65, 0.1, ALPHA)
 
     def set_validate(self, validator=None):
         if validator:
-            self.connect('changed', self.__color, validator)
+            self.connect('changed', self._set_color, validator)
+
+    def _default_color(self) -> Gdk.RGBA:
+        # Don't use our *own* context here if possible else it changes as we mutate...
+        # Don't cache, as theme switching is async
+        parent = self.get_parent()
+        ctx = parent.get_style_context() if parent else self.get_style_context()
+        col = ctx.get_color(Gtk.StateType.NORMAL)
+        return col
+
+    def _set_color(self, _widget, validator):
+        def mix(src: Gdk.RGBA, dest: Gdk.RGBA, ratio: float):
+            ratio = min(1.0, max(0.0, ratio))
+            inv = 1.0 - ratio
+            return Gdk.RGBA(inv * src.red + ratio * dest.red,
+                            inv * src.green + ratio * dest.green,
+                            inv * src.blue + ratio * dest.blue,
+                            inv * src.alpha + ratio * dest.alpha)
```

Comment:
I think this would be better in some kind of Gdk / color utils module.

Reply:
Done!